Content.Server/GameObjects/Components/Items/Storage/SingleItemStorageComponent.cs

```diff
@@ -0,0 +1,93 @@
+﻿#nullable enable
+
+using System.Collections.Generic;
+using Content.Shared.GameObjects.Components;
+using Robust.Server.GameObjects;
+using Robust.Server.GameObjects.Components.Container;
+using Robust.Server.Interfaces.GameObjects;
+using Robust.Shared.GameObjects;
+using Robust.Shared.GameObjects.ComponentDependencies;
+using Robust.Shared.Interfaces.GameObjects;
+using Robust.Shared.Log;
+using Robust.Shared.Serialization;
+
+namespace Content.Server.GameObjects.Components.Items.Storage
+{
+    /// <summary>
+    /// Storage that spawns and counts a single item.
+    /// Usually used for things like matchboxes, cigarette packs,
+    /// cigar cases etc.
+    /// </summary>
+    [RegisterComponent]
+    public class SingleItemStorageComponent : Component, IMapInit
+    {
+        private string? _prototypeName;
+        private int _amount;
+
+        /// <summary>
+        /// Storage component that we can use for inserting and counting elements.
+        /// </summary>
+        [ComponentDependency] private readonly ServerStorageComponent? _storage = default;
+        /// <summary>
+        /// Single item storage component usually have an attached StackedVisualizer.
+        /// </summary>
+        [ComponentDependency] private readonly AppearanceComponent? _appearanceComponent = default;
+
+        public override string Name => "SingleItemStorage";
```

Comment:
So if I understand correctly the point of this class is to:
A) Show an appearance update based on specific prototypes going in
B) Pre-fill that prototype where possible.

Based on this I have a couple comments:
A) You shouldn't have a reliance on prototypes for stuff. Spawning it is obviously okay but checking appearance based on prototype should really use a tag or component instead. This means you could have 5 different cigarette types and they can all affect the appearance.
B) It would be useful if we could defer spawning for these given a cigarette pack is now 11 entities which isn't cheap to spawn though I guess storage doesn't support this right? Which is a shame as it'd be a big win.
C) Maybe it's just me but SingleItem implies something like a ContainerSlot rather than a single type. The other part is that it doesn't even restrict the types of entities that goes in so it's not even really a single "type" it's just what updates the appearance.

With that being said it seems more advantageous to just do the pre-spawning on ServerStorageComponent and then you can move all of this functionality to ClientStorageComponent instead maybe?

Reply:
> A) You shouldn't have a reliance on prototypes for stuff.

So no `[ComponentDependency]`? Ok will note.

> B) It would be useful if we could defer spawning

Valid point, I wanted to find disable counting for the first N item, since no point in visualizing initial insertion of items.

> C) Maybe it's just me but SingleItem implies something like a ContainerSlot rather than a single type. The other part is that it doesn't even restrict the types of entities

The issue is, I can make SingleItem more restrictive, but for now, I wanted to keep simple. As for your suggestion. It occured to me, but then what if I stuff my cig pack with contraband? Does it look like bunch of cigars or few cigars and contraband? I opted for the latter, because if I stuff my cig pack with seven chips it should not IMO look like I have six cigarettes.